prow/templates/configtree.go

```diff
@@ -0,0 +1,83 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package templates
+
+// T is the struct holding the plugin configuration.
+type T struct{}
+
+// ConfigTree specifies the global generic config for a plugin.
+type ConfigTree struct {
+	T
+	Orgs map[string]Org `json:"orgs,omitempty"`
+}
+
+// GetOrg returns the org config after merging in any global config.
+func (t ConfigTree) GetOrg(name string) *Org {
+	o, ok := t.Orgs[name]
+	if ok {
+		o.Approve = t.Approve.Apply(o.Approve)
+	} else {
+		o.Approve = t.Approve
+	}
+	return &o
+}
+
+// Org holds the default config for an entire org, as well as any repo overrides.
+type Org struct {
+	T
+	Repos map[string]Repo `json:"repos,omitempty"`
+}
+
+// GetRepo returns the repo config after merging in any org policies.
+func (o Org) GetRepo(name string) *Repo {
+	r, ok := o.Repos[name]
+	if ok {
+		r.Approve = o.Approve.Apply(r.Approve)
+	} else {
+		r.Approve = o.Approve
+	}
+	return &r
+}
+
+// Repo holds the default config for all branches in a repo, as well as specific branch overrides.
+type Repo struct {
+	T
+	Branches map[string]Branch `json:"branches,omitempty"`
+}
+
+// GetBranch returns the branch config after merging in any repo policies.
+func (r Repo) GetBranch(name string) *Branch {
+	b, ok := r.Branches[name]
+	if ok {
+		b.Approve = r.Approve.Apply(b.Approve)
+	} else {
+		b.Approve = r.Approve
+	}
+	return &b
+}
+
+// Branch holds config for a particular branch.
+type Branch struct {
+	T
+}
+
+// GetLeafConfig returns the plugin configuration for a given org/repo/branch.
+//
+// Handles merging any policies defined at repo/org/global levels into the branch policy.
+func (t *ConfigTree) GetLeafConfig(org, repo, branch string) T {
+	return t.GetOrg(org).GetRepo(repo).GetBranch(branch).Approve
```

Comment:
`.T`?

Reply:
Please ignore... It's just that at the moment I cannot make it work for the struct elements.
Of course it should be T but doesn't yet work.
(I did push to transfer the work to my home computer)

Reply:
You can try yourself, change to T and run bazel build //...

Reply:
https://github.com/google/gvisor/blob/master/tools/go_generics/generics.go#L85
> // Unfortunately, go_generics does not handle anonymous fields with renamed types.

Reply:
By anonymous do they mean struct embedding?

Reply:
This is how I understand it, and it explains why it didn't work... Now we should decide if it's really cumbersome to add config everywhere.
BTW, looks like they want to remove embedding in golang 2 (but add generics)...